api/users/serializers.py

```diff
@@ -297,3 +301,101 @@ class UserAccountDeactivateSerializer(BaseAPISerializer):
+        except TwoFactorUserSettings.DoesNotExist:
+            return False
+
+    def get_subscribe_osf_general_email(self, obj):
+        return obj.mailchimp_mailing_lists.get(MAILCHIMP_GENERAL_LIST, False)
+
+    def get_subscribe_osf_help_email(self, obj):
+        return obj.osf_mailing_lists.get(OSF_HELP_LIST, False)
+
+    links = LinksField({
+        'self': 'get_absolute_url'
+    })
+
+    def get_absolute_url(self, obj):
+        return absolute_reverse(
+            'users:user_settings',
+            kwargs={
+                'user_id': obj._id,
+                'version': self.context['request'].parser_context['kwargs']['version']
+            }
+        )
+
+    class Meta:
+        type_ = 'user_settings'
+
+
+class UserSettingsUpdateSerializer(UserSettingsSerializer):
+    id = IDField(source='_id', required=True)
+    two_factor_enabled = ser.BooleanField(write_only=True, required=False)
+    two_factor_verification = ser.IntegerField(write_only=True, required=False)
+    subscribe_osf_general_email = ser.BooleanField(read_only=False, required=False)
+    subscribe_osf_help_email = ser.BooleanField(read_only=False, required=False)
+
+    # Keys represent field names values represent the human readable names stored in DB.
+    MAP_MAIL = {
+        'subscribe_osf_help_email': OSF_HELP_LIST,
+        'subscribe_osf_general_email': MAILCHIMP_GENERAL_LIST,
+    }
+
+    def update_email_preferences(self, instance, attr, value):
+        if self.MAP_MAIL[attr] == OSF_HELP_LIST:
+            update_osf_help_mails_subscription(user=instance, subscribe=value)
+        else:
+            update_mailchimp_subscription(instance, self.MAP_MAIL[attr], value)
+        instance.save()
+
+    def update_two_factor(self, instance, value, two_factor_addon):
+        if value:
+            if not two_factor_addon:
+                two_factor_addon = instance.get_or_add_addon('twofactor')
+                two_factor_addon.save()
+        else:
+            auth = get_user_auth(self.context['request'])
+            instance.delete_addon('twofactor', auth=auth)
+
+        return two_factor_addon
+
+    def verify_two_factor(self, instance, value, two_factor_addon):
+        if not two_factor_addon:
+            raise exceptions.ValidationError(detail='Two-factor authentication is not enabled.')
+        if two_factor_addon.verify_code(value):
+            two_factor_addon.is_confirmed = True
+        else:
+            raise exceptions.PermissionDenied(detail='The two-factor verification code you provided is invalid.')
+        two_factor_addon.save()
+
+    def to_representation(self, instance):
+        """
+        Overriding to_representation allows using different serializers for the request and response.
+        """
+        context = self.context
+        return UserSettingsSerializer(instance=instance, context=context).data
+
+    def update(self, instance, validated_data):
+
+        two_factor_addon = instance.get_addon('twofactor')
```

Comment:
Could this be moved to the twofactor functions or only execute when a twofactor attr is present? We don't want to query for the addon when the payload might only contain info to update email lists.